### PR TITLE
Fix ber scanf/printf, fix for a reverted PR.

### LINKED ISSUE #49105: [macOS-arm64] 55 System.DirectoryServices.Protocols.Tests fail on CoreCLR
See https://dev.azure.com/dnceng/public/_build/results?buildId=1021947

First few failures.
```console
    System.DirectoryServices.Protocols.Tests.PageResultRequestControlTests.Ctor_Cookie(cookie: null, expectedValue: [48, 6, 2, 2, 2, ...]) [FAIL]
      System.DirectoryServices.Protocols.BerConversionException : An error occurred during the BER conversion.
      Stack Trace:
        /_/src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/BerConverter.cs(209,0): at System.DirectoryServices.Protocols.BerConverter.Encode(String format, Object[] value)
        /_/src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs(599,0): at System.DirectoryServices.Protocols.PageResultRequestControl.GetValue()
        /_/src/libraries/System.DirectoryServices.Protocols/tests/PageResultRequestControlTests.cs(72,0): at System.DirectoryServices.Protocols.Tests.PageResultRequestControlTests.Ctor_Cookie(Byte[] cookie, Byte[] expectedValue)
    System.DirectoryServices.Protocols.Tests.SecurityDescriptorFlagControlTests.Ctor_Flags(masks: Group, expectedValue: [48, 3, 2, 1, 2]) [FAIL]
      Assert.Equal() Failure
      Expected: Byte[] [48, 3, 2, 1, 2]
      Actual:   Byte[] [48, 6, 2, 4, 112, ...]
      Stack Trace:
        /_/src/libraries/System.DirectoryServices.Protocols/tests/SecurityDescriptorFlagControlTests.cs(41,0): at System.DirectoryServices.Protocols.Tests.SecurityDescriptorFlagControlTests.Ctor_Flags(SecurityMasks masks, Byte[] expectedValue)
    System.DirectoryServices.Protocols.Tests.ExtendedDNControlTests.Ctor_Default [FAIL]
      Assert.Equal() Failure
      Expected: Byte[] [48, 3, 2, 1, 0]
      Actual:   Byte[] [48, 6, 2, 4, 112, ...]
      Stack Trace:
        /_/src/libraries/System.DirectoryServices.Protocols/tests/ExtendedDNControlTests.cs(23,0): at System.DirectoryServices.Protocols.Tests.ExtendedDNControlTests.Ctor_Default()
    System.DirectoryServices.Protocols.Tests.VlvRequestControlTests.Ctor_BeforeCount_AfterCount_ByteArrayTarget(beforeCount: 0, afterCount: 0, target: null, expectedValue: [48, 14, 2, 1, 0, ...]) [FAIL]
      Assert.Equal() Failure
      Expected: Byte[] [48, 14, 2, 1, 0, ...]
      Actual:   Byte[] [48, 30, 2, 4, 112, ...]
      Stack Trace:
        /_/src/libraries/System.DirectoryServices.Protocols/tests/VlvRequestControlTests.cs(102,0): at System.DirectoryServices.Protocols.Tests.VlvRequestControlTests.Ctor_BeforeCount_AfterCount_ByteArrayTarget(Int32 beforeCount, Int32 afterCount, Byte[] target, Byte[] expectedValue)
    System.DirectoryServices.Protocols.Tests.BerConverterTests.Encode_Objects_ReturnsExpected(format: "tetie", values: [128, 0, 133, 2, 3], expected: [128, 1, 0, 133, 1, ...]) [FAIL]
      Exception of type 'Xunit.Sdk.AssertActualExpectedException' was thrown.
      Expected: 128, 1, 0, 133, 1, 2, 10, 1, 3
      Actual:   1, 112, 0, 0, 116, 4, 112, 0, 0, 101, 1, 112, 0, 0, 116, 4, 112, 0, 0, 105, 10, 4, 112, 0, 0, 101
      Stack Trace:
        /_/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs(370,0): at System.AssertExtensions.Equal[T](T[] expected, T[] actual)
        /_/src/libraries/System.DirectoryServices.Protocols/tests/BerConverterTests.cs(49,0): at System.DirectoryServices.Protocols.Tests.BerConverterTests.Encode_Objects_ReturnsExpected(String format, Object[] values, Byte[] expected)
```

/cc @sandreenko These look a bit like ABI issues.

### PR DESCRIPTION
**Pull request by sandreenko**
https://github.com/dotnet/runtime/pull/51205 was [reverted ](https://github.com/dotnet/runtime/pull/52038)because of the [failures ](https://github.com/dotnet/runtime/issues/49105#issuecomment-828897605)on arm64 Linux.

@jkoritzinsky, @AaronRobinsonMSFT PTAL at the last commit here. Thanks for pointing me to https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#cc-long, I have tried to use `CULong` but it is supported only by 6.0, so I would have to delete `netcoreapp2.0` and `netstandard2.0` support
from https://github.com/dotnet/runtime/blob/d2fba8fdc44f5ed7a4676c219c626e36a3fa2998/src/libraries/System.DirectoryServices.Protocols/src/System.DirectoryServices.Protocols.csproj#L5
and https://github.com/dotnet/runtime/blob/d2fba8fdc44f5ed7a4676c219c626e36a3fa2998/src/libraries/System.DirectoryServices.AccountManagement/src/System.DirectoryServices.AccountManagement.csproj#L6
and I guess it is not desired?

The fix was to use nuint instead of unit so the default tag value is passed as `0xffffffffffffffff` on Linux. This parameter is not used for Windows, so we don't care about it there.

Fixes https://github.com/dotnet/runtime/issues/49105.

### CODE REVIEW

**Comment by sandreenko** [comment]
The tests involving the new code have passed on windows x64/linux x64/linux arm64/MacOS x64/MacOS arm64.
The failures are unrelated.

**Comment by sandreenko** [comment]
I would like to merge this PR this week if possible, could somebody please take a look? @AaronRobinsonMSFT, @jkoritzinsky

**Review by AaronRobinsonMSFT — APPROVED** [review]
I think there is a place where `tag` is used when it should be passed as a 32-bit integer not 64-bit. Overall I don't see anything of concern.

File: src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/Interop/BerPal.Windows.cs [hunk 1]
```diff
@@ -15,19 +15,19 @@ internal static class BerPal
 
         internal static int FlattenBerElement(SafeBerHandle berElement, ref IntPtr flattenptr) => Interop.Ldap.ber_flatten(berElement, ref flattenptr);
 
-        internal static int PrintBerArray(SafeBerHandle berElement, string format, IntPtr value, int tag) => Interop.Ldap.ber_printf(berElement, format, __arglist(value));
+        internal static int PrintBerArray(SafeBerHandle berElement, string format, IntPtr value, nuint tag) => Interop.Ldap.ber_printf(berElement, format, __arglist(value));
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/Interop/BerPal.Windows.cs:18]: Can I have several args named "_", like:
```
public static int ber_printf_tag(SafeBerHandle _, string _, nuint _)
```
?

File: src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/Interop/BerPal.Windows.cs [hunk 2]
```diff
@@ -15,21 +15,25 @@ internal static class BerPal
...
+        internal static int PrintByteArray(SafeBerHandle berElement, string format, HGlobalMemHandle value, uint length, nuint tag) => Interop.Ldap.ber_printf(berElement, format, __arglist(value, length));
 
-        internal static int PrintEmptyArgument(SafeBerHandle berElement, string format) => Interop.Ldap.ber_printf_emptyarg(berElement, format);
+        internal static int PrintEmptyArgument(SafeBerHandle berElement, string format, nuint tag) => Interop.Ldap.ber_printf(berElement, format, __arglist());
 
-        internal static int PrintInt(SafeBerHandle berElement, string format, int value) => Interop.Ldap.ber_printf_int(berElement, format, value);
+        internal static int PrintInt(SafeBerHandle berElement, string format, int value, nuint tag) => Interop.Ldap.ber_printf(berElement, format, __arglist(value));
 
-        internal static int ScanNext(SafeBerHandle berElement, string format) => Interop.Ldap.ber_scanf(berElement, format);
+        internal static int PrintTag(SafeBerHandle berElement, string format, nuint tag) => Interop.Ldap.ber_printf(berElement, format, __arglist(tag));
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/Interop/BerPal.Windows.cs:26]: you are right, fixed, don't understand why [x86 Libraries](https://dev.azure.com/dnceng/public/_build/results?buildId=1118211&view=logs&j=3258e425-c037-59ba-e3ca-218fa6e3ed33&t=f51f65e4-b70d-5e1f-1a4a-9bd196f052de) tests passed with this.

**Comment by sandreenko** [comment]
runtime-libraries-coreclr outerloop had [32 failures](https://dev.azure.com/dnceng/public/_build/results?buildId=1127997&view=ms.vss-test-web.build-test-results-tab), runtime-libraries-coreclr outerloop-osx had [8 failures](https://dev.azure.com/dnceng/public/_build/results?buildId=1127999&view=ms.vss-test-web.build-test-results-tab). They all were marked as new but looked unrelated but I did not see some of them in the previous so I decided to double-check and triggered the main testing and it showed the same results without my changes runtime-libraries-coreclr outerloop [38 failures](https://dev.azure.com/dnceng/public/_build/results?buildId=1128611&view=ms.vss-test-web.build-test-results-tab), runtime-libraries-coreclr outerloop-osx [6 failures](https://dev.azure.com/dnceng/public/_build/results?buildId=1128610&view=ms.vss-test-web.build-test-results-tab) and the set was ~the same.
So I think the failures are not related.

**Pull request merged**